spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala

```diff
@@ -104,7 +103,7 @@ case class RewriteMergeInto(spark: SparkSession) extends Rule[LogicalPlan] with
       case MergeIntoTable(target: DataSourceV2Relation, source: LogicalPlan, cond, matchedActions, notMatchedActions)
           if notMatchedActions.isEmpty =>
 
-        val (mergeBuilder, targetTableScan) = buildDynamicFilterTargetScan(target, source, cond, matchedActions)
+        val mergeBuilder = target.table.asMergeable.newMergeBuilder("merge", newWriteInfo(target.schema))
```

Comment:
I think we better move the merge builder creation from `buildDynamicFilterTargetScan`.